Turning segfaults into exceptions with high probability

nh2 opened the pull request:
The IOHK devops team has complained that incorrect usage of this library can result in hard crashes (segmentation faults) that can be hard to debug.

The changes in this PR turn most of the errors that arise from incorrect usage (such as calling `close` twice) from segmentation faults into Haskell exceptions with a helpful error message.

Further:

* I've added a bunch of tests that show how the various ways to open and close DBs fail
* I've added lots of Haddocks to the `open*`/`close` functions so that callers are aware how they fail
* I have added `HasCallStack` constraints to many functions that can fail (such as `close`), so that when an exception arises, users can easily see where it `close` was used incorrectly
* I've added `withDB` as a simple bracket-style function; ideally most users could use that and never call `close` manually, like for example:

```
main = do
  ...
  withDB path options $ \db -> do
    ... all program code, never calling `close` ...
```

nh2 committed:
tests: Fix directories being left over in system temp dir

nh2 committed:
tests: Change misleading type signature.

The `MonadResource` wasn't actually used, wrongly suggesting that
the DB would be auto-closed after usage.

nh2 committed:
tests: Don't forget to close DB

nh2 committed:
tests: Add positive-regresion check for LOCK check behaviour

nh2 committed:
tests: Add test that segfaults. Makes the tests fail.

nh2 committed:
Add comment about async-exception saety of `open`

nh2 committed:
Add comment about double-close segfault

nh2 committed:
Turn incorrect usage (use-after-close) from SEGFAULT into clean errors in most cases.

See the comment added to the `DB` type for details.

Segfaults are very nasty to track down and don't report an error message
that end users can report to developers, so any case where we can turn
a segfault into a clean error message is an improvement.

Also add `withDB` function for clean scoped opening, and add `HasCallStack`
constraints on functions that can error so that callers can more easily spot
where in their code the mistake is if we can spot it.

This also fixes the tests again that were segfaulting as introduced
in the previous commit.

domenkozar commented:
Fixes https://github.com/serokell/rocksdb-haskell/issues/9

domenkozar commented:
@gromakovsky @neongreen

nh2 committed:
Remove openAutoclose again.

I realise I cannot add that without moving a lot of code into C
(currently when you don't call `finalize` explicitly, it doesn't
free the `Options'`, and I can't make it free the `Options'` when
a `DB` is freed unless I move it to C).

This was easily found by valgrind 3.13 (3.11 crashed).

I also don't like that its usefulness is limited by that global table
RocksDB has, so one can't really use `openAutoclose` without `finalize`.

Also fix a test to free DB eventually.
All tests are now valgrind-memcheck-clean.

domenkozar commented:
Appveyor:

```
Failures:
  tests\tests.hs:62: 
  1) Basic DB Functionality does weird global singleton string matching stuff for double-locking warnings
       uncaught exception: IOException of type PermissionDenied (removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "/tmp/tmpdir\\LOCK": permission denied (The process cannot access the file because it is being used by another process.))
Randomized with seed 390801961
Finished in 5.2562 seconds
```

nh2 commented:
> `IOException of type PermissionDenied (removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "/tmp/tmpdir\\LOCK": permission denied (The process cannot access the file because it is being used by another process.))`

I guess that makes sense on Windows. I should `#ifdef` that test out on that platform.

nh2 committed:
tests: Remove string matching test on Windows.

Should fix this failure on AppVeyor:

Failures:
  tests\tests.hs:62:
  1) Basic DB Functionality does weird global singleton string matching stuff for double-locking warnings
       uncaught exception: IOException of type PermissionDenied (removeDirectoryRecursive:removeContentsRecursive:removePathRecursive:DeleteFile "/tmp/tmpdir\\LOCK": permission denied (The process cannot access the file because it is being used by another process.))
Randomized with seed 390801961
Finished in 5.2562 seconds

We also have to remove the multiline string because CPP doesn't work
with multiline strings.

nh2 committed:
tests: Use `runResourceTChecked`.

Also update comments accordingly.

gromakovsky commented on src/Database/RocksDB/Base.hs:
```diff
@@ -154,50 +169,75 @@ open path opts = liftIO $ bracketOnError initialize finalize mkDB
+close :: (MonadIO m, HasCallStack) => DB -> m ()
+close db@DB{ db_ptr, dbOptions } = liftIO $ do
+    ensureOpenAndClose db
+    c_rocksdb_close db_ptr `finally` freeOpts dbOptions
```
What if there is an async exception after `ensureOpenAndClose` and before this line? AFAIU, we will set status to `Closed`, but won't actually close the DB.
But on the other hand this function is intended to be used as a release action in `bracket` or anything similar, so it shouldn't be a problem I suppose.

nh2 replied:
@gromakovsky That is correct.

None of the "plain" functions in this module are async-exception-safe; the same comment [as I wrote on `open`](https://github.com/serokell/rocksdb-haskell/pull/10/files/fc69de928e8fafb529e571320f88f3f6f4be1978#diff-56244378bea7e31bbe12cb39da7b646cR145) applies. It is not clear to me if the library authors had it this way by design (it might be that they planned for all plain functions to be intended to only be used in call-sites that themselves use `bracket`) or by accident, so I haven't made `close` use `mask_ $ do` so far.

I should probably copy-paste that comment here, though it might possibly have to be copied to a lot more functions (but I haven't analysed in detail which), so until now I have instead Haddocked the functions that _are_ async-exception-safe.

Only `withDB` and `openBracket` are `async-exception-safe`.

chrisdone replied:
What about changing 

```haskell
data DB = DB RocksDBPtr Options'
```

to

```haskell
data DB = DB (IORef (Maybe (RocksDBPtr, Options'))) 
```

And the `close`  can be implemented as:

```haskell
close (DB m_ref) = liftIO $ do
    atomicModifyIORef'
        m_ref
        (\m ->
             fmap
                 (Nothing, )
                 (case m of
                      Nothing -> pure ()
                      Just (db_ptr, opts_ptr) -> do
                          c_rocksdb_close db_ptr 
                          freeOpts opts_ptr))
```

This way:

1. You can't double close: in a single thread (or multiple threads; IORef's `atomicModifyIORef'` guarantees ordering).
2. Any other operation that needs a db_ptr (read/put/modify) can simply throw an exception if the m_ref contains a `Nothing`, complaining that the database has closed.
3. IORef operations are "guaranteed not to be interruptible", so you can't get an async exception while performing them.

gromakovsky reviewed (changes requested):
1. Can you please update CHANGELOG.md and increase package version?
2. I wish there was an easier way to statically check that `ensureOpen` is present in all places where we use `db_ptr`. What do you think about adding `checkedDbPtr :: DB -> IO RocksDBPtr` (feel free to use better naming, it's just an arbitrary name) which does `ensureOpen` internally and then using `checkedDbPtr` almost eveywhere? Then we can find all usages of `db_ptr` and check they are all correct. There should be only a few of them. Currently number of `db_ptr` usages is `O(number of rocksdb functions)`. We have bindings for a small number of rocksdb functions, so it's not too hard to inspect all `db_ptr`, but if later we add bindings for whole rocksdb, it will be too much.

Anyway, thanks for this patch, the second point is just polishing (and maybe there are reasons why it's not a good idea), so I will approve regardless of whether it will be done, but please do (1).

gromakovsky commented on tests/tests.hs:
```diff
@@ -1,44 +1,145 @@
         put db def "zzz" "zzz"
-        get db def "zzz"
+        val <- get db def "zzz"
+        return val
```
Btw, why did you change:
```
get db def "zzz"
```
to
```
        val <- get db def "zzz"
        return val
```

Do you think the latter is more readable? 🤔

chrisdone replied:
(It's likely that he was printing or debugging `val` and then didn't come back and return it to the old code.)

nh2 replied:
This actually comes from [here](https://github.com/serokell/rocksdb-haskell/pull/10/commits/fc69de928e8fafb529e571320f88f3f6f4be1978#diff-bfd1c1e5fa80be47287fe256c3cbb01aL59):

I changed to bin the value so that I could insert a `close` in between, and then removed the close again when the `resourcet` patch allowed me to do so:

```diff
         val <- get db def "zzz"
-        close db -- TODO remove; this should make the test fail, but it doesn't because errors in finalizers run by runResourceT don't seem to bubble up
         return val
``` 

We could change it back to just `get db def "zzz"`, I have no opinion on which one is more readable.

gromakovsky commented on tests/tests.hs:
```diff
@@ -1,44 +1,145 @@
+    it "does weird global singleton string matching stuff for double-locking warnings" $ do
+      -- This test will fail when RocksDB fixes this issue
+      -- (see https://stackoverflow.com/questions/37310588/rocksdb-io-error-lock-no-locks-available#comment83145041_37312033).
+      -- It exists so that we get notified when that fixing happens.
```
Is there anything we should do on our end when it happens? :)

nh2 replied:
That depends on what they do instead. If they do what sqlite would probably do, then no. If they come up with some other way, you might want to inspect whether it's compatible with your way of using it.

chrisdone committed:
Update snoyberg/conduit reference in stack.yaml

CMCDragonkai commented:
Is this still being worked on?

nh2 commented:
> Is this still being worked on?

@CMCDragonkai Not by us currently; FP Complete are no longer for contracting for IOHK.

We also made https://github.com/chrisdone/rocksdb-haskell-ng as a from-scratch, safe-from-the-beginning replacement for `rocksdb-haskell` for IOHK after this PR.

If you need this issue fixed, would benefit from `rocksdb-haskell-ng`, or any other Haskell RocksDB help, let us know!

CMCDragonkai commented:
@nh2 Why is that repo read-only/archived? Is there an active repo where this stuff is still being developed.

nh2 commented:
> Why is that repo read-only/archived?
> Is there an active repo where this stuff is still being developed.

@CMCDragonkai FP Complete made `rocksdb-haskell-ng` for IOHK to fix the mentioned issues. It seems IOHK is still using their fork that's a couple commits behind our archived repo (see [here](https://github.com/input-output-hk/cardano-sl/blob/2492992e65fe33ffd0cd93d735097f12b5917945/nix/.stack.nix/rocksdb-haskell-ng.nix#L54), and the [list of forks](https://github.com/chrisdone/rocksdb-haskell-ng/network/members)), but we do not currently have any clients that want us to maintain it so we archived it.

We'd certainly be happy to revive it or improve it for you if that'd be useful for you. Just let me know!
